### PR TITLE
test: Add lots more unit tests, change application from using stdout

### LINKED ISSUE #6: Unit testing
Ensure that all console output is to dependency injected output interfaces rather than using log.Fmt() etc. This will allow us to unit test the output of these functions.

Produce unit tests as examples for existing functions, mocking out the requests to the API.

### PR DESCRIPTION
**Pull request by JakeMakesStuff**
Unit tests all commands in the application which rely on a response and user input

Fixes #6

### CODE REVIEW

File: cmd/katapult/config_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+		apiURL: "test",
+	},
+}
+
+func TestConfig(t *testing.T) {
+	conf, err := config.New()
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, v := range mockConfigSuite {
```
Comment (strideynet) [inline cmd/katapult/config_test.go:44]: Use Subtests here https://blog.golang.org/subtests

We tend to stick to using the variable name ``tt`` for the current test case executing.

File: cmd/katapult/networks_test.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+			Country: &core.Country{
+				ID:   "pog",
+				Name: "Pogland",
+			},
+		},
+	},
+}
+
+func mockNetworksServer() (net.Listener, error) {
+	ln, err := net.Listen("tcp", "127.0.0.1:7654")
```
Comment (strideynet) [inline cmd/katapult/networks_test.go:78]: If you are going to use local http server rather than mocking with an interface, you should look at https://golang.org/pkg/net/http/httptest/#NewServer . This gives you a random port, which means we can run the tests concurrently.

You can see some prior art here on one of my repos: https://github.com/zmb3/spotify/blob/master/spotify_test.go#L12

File: cmd/katapult/networks_test.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+		}
+		writer.Header().Add("Content-Type", "application/json")
+		writer.WriteHeader(200)
+		_, _ = writer.Write(b)
+	})
+	go func() { _ = http.Serve(ln, mux) }()
+	return ln, nil
+}
+
+var expectedNetworkResults = []struct {
```
Comment (strideynet) [inline cmd/katapult/networks_test.go:114]: Nice test table. Usually I'd expect to see it inside of the test function itself, and usually named "tests".

File: cmd/katapult/networks_test.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		t.Fatal(err)
+	}
+	defer ln.Close()
+	client, err := newClient(&config.Config{APIURL: "http://127.0.0.1:7654", APIKey: "a"})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Validate the get command.
+	for _, v := range expectedNetworkResults {
```
Comment (strideynet) [inline cmd/katapult/networks_test.go:152]: Again, would be nice to see ``tt`` as the variable here and a subtest introduced.

File: cmd/katapult/networks_test.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+ - Pognet 3 [pognet3]
+ - Pognet 4 [pognet4]
+`,
+	},
+}
+
+func TestNetworkList(t *testing.T) {
+	// Create the mock server.
+	ln, err := mockNetworksServer()
+	if err != nil {
```
Comment (strideynet) [inline cmd/katapult/networks_test.go:142]: What can be quite nice, is passing ``t`` into your mock helpers and letting them call t.Fatal() internally. Just means you don't have to do the if err boilerplate.

**Review by strideynet — CHANGES_REQUESTED** [review]

File: cmd/katapult/config_test.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+func TestConfig(t *testing.T) {
+	conf, err := config.New()
+	if err != nil {
+		t.Fatal(err)
+	}
+	tests := []struct {
+		apiKey string
+		apiURL string
+	}{
```
Comment (strideynet) [inline cmd/katapult/config_test.go:25]: Usually I'd expect a test table to include some sort of expected output within the table, so its really clear what the expected behaviour is.

File: cmd/katapult/mock_api_client.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+	},
+	{
+		ID:        "testing",
+		Name:      "testing, testing, 123",
+		SubDomain: "test",
+	},
+}
+
+// Used to mock the API client for the purpose of unit tests.
+type mockAPIClient struct{}
```
Comment (strideynet) [inline cmd/katapult/mock_api_client.go:113]: @jimeh I'd be curious on your thoughts on mocking like this for the CLI, rather than using an interface style mock?

**Review by strideynet — CHANGES_REQUESTED** [review]

File: cmd/katapult/client_test.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+	// Test invalid base URL.
+	_, err = newClient(&config.Config{
+		APIKey: "test",
+		APIURL: "this is a test",
+	})
+	if err == nil {
+		t.Fatal("expected error")
+	}
+}
```
Comment (jimeh) [inline cmd/katapult/client_test.go:48]: I think these two tests could benefit from being turned into a single `Test_newClient` tabular test that has test cases for setting each of the available attributes, and also one for setting all of the attributes.

Also I would recommend using [`testify/assert`](https://github.com/stretchr/testify#assert-package) for assertions to make things cleaner and easier. We tend to use it in pretty much all of our Go projects already :)

File: cmd/katapult/config_test.go (hunk 6 above)
Comment (jimeh) [reply cmd/katapult/config_test.go:25]: Agreed. I think a `name` field with a descriptive text for each test case would help as well. Cause right now it's rather difficult to quickly grasp what this test is actually testing.

File: cmd/katapult/mock_api_client.go (hunk 7 above)
Comment (jimeh) [reply cmd/katapult/mock_api_client.go:113]: I think I would much rather prefer interface mocks via gomock/mockgen or something similar.

While this kind of straight up fake implementation of the `core.RequestMaker` interface works, I think it will become quite tedious and cumbersome to maintain as the CLI tool grows and makes more requests to Katapult's API.

Hence we should probably have an interface that covers the bits of `core.DataCentersClient` that we use, and generate a mock for it via mockgen. Then have each test-case setup the mock with expected calls and return values as needed. I have an example of this in a private repo, so please shout if that might be helpful to see.

**Review by jimeh — CHANGES_REQUESTED** [review]

File: cmd/katapult/mock_api_client.go (hunk 7 above)
Comment (strideynet) [reply cmd/katapult/mock_api_client.go:113]: Yes, I think I agree having an interface probably makes more sense. I'd like to avoid using mockgen and probably just hand code the mock though, it doesn't take long and is a bit more understandable.

Comment (jimeh) [reply cmd/katapult/mock_api_client.go:113]: I would say it depends on how varied the behavior of the mock needs to be, which depends on both how the interface is used, and what is being tested. At a certain point of complexity, the custom mock would need to be flexible enough that it's basically just providing the same feature set as gomock, but with a different/custom API to use it. At that point I'm in favor of just using gomock/mockgen :)

Personally though I'm somewhat undecided, but probably more in favor of just using gomock for consistency, unless it's something very simple which I know will only ever be used in one or two ways. That said, I haven't used gomock all that much yet, or at least not enough to have a bunch of paper cuts from it making me cautious of using it... lol

My primary goal with these kind of things are to try and keep the mock/fake data/logic—however it's defined—tightly coupled to the test case code itself. This avoids long if/else chains elsewhere that cover all manner of test cases across various tests, and it makes it easier to look at a single test case and understand what it does. That said, if lots of test cases setup mocks the same way, I'm happy for helper functions to be used to reduce code duplication, as long as the helper functions have obvious names so it's still reasonably clear what's going on.

File: cmd/katapult/client_test.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+			apiURL: "https://example.com",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c, err := newClient(&config.Config{
+				APIKey: tt.apiKey,
+				APIURL: tt.apiURL,
+			})
+			if err != nil {
```
Comment (strideynet) [inline cmd/katapult/client_test.go:36]: require.NoError(t, err)

File: cmd/katapult/config.go [hunk 10]
```diff
@@ -15,14 +14,16 @@ func configCommand(conf *config.Config) *cobra.Command {
 		Use:   "config",
 		Short: "Print configuration",
 		Long:  `Print parsed configuration in YAML format.`,
-		Run: func(cmd *cobra.Command, args []string) {
+		RunE: func(cmd *cobra.Command, args []string) error {
 			bs, err := yaml.Marshal(conf.AllSettings())
 			if err != nil {
-				log.Fatalf("unable to marshal config to YAML: %v", err)
+				return fmt.Errorf("unable to marshal config to YAML: %v", err)
```
Comment (strideynet) [inline cmd/katapult/config.go:20]: Use %w when wrapping errors :)

**Review by strideynet — CHANGES_REQUESTED** [review]

File: cmd/katapult/mocks/katapult_request_matcher.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+	if !ok {
+		// This should never happen.
+		return false
+	}
+	if req.URL.Path != k.Path {
+		// This is for the wrong path.
+		return false
+	}
+	q := req.URL.Query()
+	for _, x := range k.ExpectedParams {
```
Comment (strideynet) [inline cmd/katapult/mocks/katapult_request_matcher.go:25]: Would this catch extra params ? Or is that not a concern ?

Comment (JakeMakesStuff) [reply cmd/katapult/mocks/katapult_request_matcher.go:25]: Not a concern. Extra params would be on a library side and would not influence the result.

**Comment by strideynet** [comment]
It might make more sense here to create interfaces that match what the commands need e.g

```
type loadBalancerController interface {
	List(ctx context.Context, org core.OrganizationRef, opts *core.ListOptions) ([]*core.LoadBalancer, *katapult.Response, error)
	Delete(ctx context.Context, lb core.LoadBalancerRef) (*core.LoadBalancer, *katapult.Response, error)
	Update(ctx context.Context, lb core.LoadBalancerRef, args *core.LoadBalancerUpdateArguments) (*core.LoadBalancer, *katapult.Response, error)
	Create(ctx context.Context, org core.OrganizationRef, args *core.LoadBalancerCreateArguments) (*core.LoadBalancer, *katapult.Response, error)
}
```

and mock those calls out using go-mock. That way we avoid relying on any details of the client library implementation. Thoughts @jimeh  ?

File: cmd/katapult/data_centers_test.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+			return nil, fmt.Errorf("unknown datacentre")
+		}).
+		Times(len(tests))
+
+	cmd := dataCentersCmd(mock)
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			stdout := &bytes.Buffer{}
+			cmd.SetOut(stdout)
+			cmd.SetErr(&bytes.Buffer{}) // Ignore stderr, this is just testing the command framework, not the error.
```
Comment (strideynet) [inline cmd/katapult/data_centers_test.go:119]: We probably want to ensure nothing is written to the stderr, if we aren't expecting it.

File: cmd/katapult/client_test.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/stretchr/testify/require"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/krystal/go-katapult"
+
+	"github.com/krystal/katapult-cli/config"
+)
+
+func TestNewClient(t *testing.T) {
```
Comment (jimeh) [inline cmd/katapult/client_test.go:15]: Nitpick: I believe this test should be called `Test_newClient` as the function it's testing is called `newClient`.

File: cmd/katapult/client_test.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+			apiURL: "https://example.com",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c, err := newClient(&config.Config{
+				APIKey: tt.apiKey,
+				APIURL: tt.apiURL,
+			})
+			require.NoError(t, err)
```
Comment (jimeh) [inline cmd/katapult/client_test.go:38]: As `newClient` can return an error, and has a couple of error specific cases of logic to handle errors, I think we should probably have some test cases that cover the possible error conditions.

File: cmd/katapult/data_centers_test.go [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+		dc     string
+		wants  string
+		stderr string
+		err    string
+	}{
+		{
+			name: "display POG1",
+			args: []string{"get", "POG1"},
+			wants: `hello (Hello World!) [POG1] / Pogland
+`,
```
Comment (jimeh) [inline cmd/katapult/data_centers_test.go:77]: This is a personal thing, so feel to disagree :)... I just dislike how raw strings mess with indentation, so for a single thing like this, I prefer a double quoted string with `\n` at the end to ensure the string ends with a line feed.

```suggestion
			wants: "hello (Hello World!) [POG1] / Pogland\n",
```

File: cmd/katapult/config_test.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+		t.Fatal(err)
+	}
+	tests := []struct {
+		name string
+
+		apiKey string
+		apiURL string
+	}{
+		{
+			name:   "Check that the config command displays blank values properly",
```
Comment (jimeh) [inline cmd/katapult/config_test.go:29]: Nitpick: I think these test case names can be shorted quite a bit as there's lots of repeated words. This one for example I think is enough if it was just `"empty values"`.

File: cmd/katapult/data_centers_test.go [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+			Name: "United Kingdom",
+		},
+	},
+}
+
+const stdoutDcList = ` - hello (Hello World!) [POG1] / Pogland
+ - hello (Hello World!) [GB1] / United Kingdom
+`
+
+func TestDataCenters_List(t *testing.T) {
```
Comment (jimeh) [inline cmd/katapult/data_centers_test.go:43]: I think this should probably a be a tabular test, as there's at least three different scenarios that might happen:

- Successful response with list of Data Centers.
- Successful response but list of Data Centers is empty (should never happen in reality, but probably worth a quick test case).
- Error response received from the API for some reason.

**Review by jimeh — CHANGES_REQUESTED** [review]

**Comment by jimeh** [comment]
@strideynet I agree about having custom interfaces for the specific resource clients from go-katapult that we use, and mock those.

Mocking the `core.RequestMaker` interface from go-katapult is fine, but it does couple of our tests pretty tightly to the `katapult/core` package, as changes to the internals of how requests are made in go-katapult that do no affect the external API of the package, would still break our tests in the CLI tool.

It does mean we probably need to do a bit more general setup for the commands in general, so we have a common place where the `core.DataCentersClient` object is defined and re-used from all data center related commands, so tests can simply replace it with a interface mock.

And related to that, I think it would be nice if we could run tests against a root command, so for example to get a DC, the args list passed in would be `[]string{"dc", "get", "uk-lon-01"}` or something to that effect. Basically so we're testing that not only does the command work in isolation, but also that it is accessible via the correct series of commands and sub-commands. It would require a pretty open structure though so the resource clients could still be easily replaced with mocks within tests.

File: cmd/katapult/data_centers.go [hunk 18]
```diff
@@ -39,32 +39,44 @@ func dataCentersCmd(client *katapult.Client) *cobra.Command {
...
 		Use:   "get",
 		Args:  cobra.ExactArgs(1),
 		Short: "Get details for a data center",
 		Long:  "Get details for a data center.",
 		RunE: func(cmd *cobra.Command, args []string) error {
-			dcClient := core.NewDataCentersClient(client)
-			dc, resp, err := dcClient.Get(cmd.Context(), core.DataCenterRef{Permalink: args[0]})
+			dc, resp, err := client.Get(cmd.Context(), core.DataCenterRef{Permalink: args[0]})
 			if err != nil {
 				if resp != nil && resp.StatusCode == 404 {
```
Comment (jimeh) [inline cmd/katapult/data_centers.go:53]: As of go-katapult 0.1.5 with improved error handling, we no longer need check the response `StatusCode` to know if a 404 not found error was encountered or not. Hence this can be simplified like so:

```suggestion
			dc, _, err := client.Get(cmd.Context(), core.DataCenterRef{Permalink: args[0]})
			if err != nil {
				if errors.Is(err, katapult.ErrNotFound) {
```

Comment (jimeh) [reply cmd/katapult/data_centers.go:53]: Also thanks to the coverage report, I noticed this not found condition is never triggered, so I think we should have a test case for when the specific data center does not exist as well.

File: cmd/katapult/networks.go [hunk 19]
```diff
@@ -34,20 +42,21 @@ func networksCmd(client *katapult.Client) *cobra.Command {
...
 			}
 
-			nets, vnets, _, err := core.NewNetworksClient(client).List(cmd.Context(), ref)
+			nets, vnets, _, err := client.List(cmd.Context(), ref)
 			if err != nil {
 				return err
 			}
 
-			fmt.Println("Networks:")
+			stdout := cmd.OutOrStdout()
```
Comment (jimeh) [inline cmd/katapult/networks.go:50]: Nitpick: As the `stdout` variable here may not always be actual STDOUT, I think calling this variable `out` or just `w` (writer) might cause slightly less confusion.

File: cmd/katapult/networks_test.go [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+	case org.SubDomain == "pog-subdomain":
+		return subdomainNetworks, nil, nil, nil
+	case org.ID == "pog-id":
+		return idNetworks, nil, nil, nil
+	default:
+		return nil, nil, nil, katapult.ErrNotFound
+	}
+}
+
+func TestNetworkList(t *testing.T) {
```
Comment (jimeh) [inline cmd/katapult/networks_test.go:91]: Nitpick: Keep test function names consistent.

```suggestion
func TestNetworks_List(t *testing.T) {
```

File: cmd/katapult/networks_test.go [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+	tests := []struct {
```
Comment (jimeh) [inline cmd/katapult/networks_test.go:92]: I think we need two additional test cases to cover everything the networks command does:

- Neither `--id` nor `--subdomain` are provided, yielding an error.
- Response includes a list of virtual networks along with networks.

File: cmd/katapult/organizations_test.go [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+ - testing, testing, 123 (test) [testing]
+`
+
+type mockOrganisationsListClient struct{}
+
+func (mockOrganisationsListClient) List(context.Context) ([]*core.Organization, *katapult.Response, error) {
+	return organizations, nil, nil
+}
+
+func TestOrganizations_List(t *testing.T) {
```
Comment (jimeh) [inline cmd/katapult/organizations_test.go:37]: I think this should be converted to a tabular test which also tests for when the API response returns an error, or an empty list of organizations, which a lot more likely to happen in the real world compared to the data centers list coming back empty :)

File: cmd/katapult/organizations.go [hunk 23]
```diff
@@ -1,16 +1,23 @@
...
 
 import (
+	"context"
 	"fmt"
 
-	"github.com/krystal/go-katapult/core"
-
 	"github.com/krystal/go-katapult"
 
+	"github.com/krystal/go-katapult/core"
```
Comment (jimeh) [inline cmd/katapult/organizations.go:9]: Nitpick: I prefer grouping external imports together into a single section without blank lines, as I think it's a bit neater, specially once you end up with a lot of imports.

**Review by jimeh — CHANGES_REQUESTED** [review]
Looking good :)

Via to code coverage I noticed a couple of minor test cases which I think we should cover, and I also had a few nitpicks. Feel free to disagree with me though... lol

File: cmd/katapult/client_test.go [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+			})
+			switch {
+			case err == nil:
+				// Ignore this. Jump out here.
+			case tt.err != "":
+				assert.Equal(t, tt.err, err.Error())
+				return
+			default:
+				t.Fatal(err)
+			}
```
Comment (jimeh) [inline cmd/katapult/client_test.go:48]: The `err == nil` case here allows this test to pass for a test case that expects an error to happen if `err` is `nil`.

Hence I suggest we make the assertion dependent purely based on the `tt.err` value, and it means we can simplify the switch statement to a single if statement:

```suggestion
			if tt.err != "" {
				require.EqualError(t, err, tt.err)
				return
			}

			require.NoError(t, err)
```

File: cmd/katapult/client_test.go [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_newClient(t *testing.T) {
+	tests := []struct {
+		name string
+
+		apiKey string
+		apiURL string
+		err    string
```
Comment (jimeh) [inline cmd/katapult/client_test.go:17]: Nitpick: I personally tend to call this field `wantErr` just to make it obvious it's the expected error.

File: cmd/katapult/data_centers_test.go [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+			err := cmd.Execute()
+			switch {
+			case err == nil:
+				// Ignore this.
+			case tt.err != "":
+				assert.Equal(t, tt.err, err.Error())
+				return
+			default:
+				t.Fatal(err)
+			}
```
Comment (jimeh) [inline cmd/katapult/data_centers_test.go:105]: I think we can simplify this switch case to a single if condition, check my comment on `Test_newClient` for reference.

File: cmd/katapult/error_utils.go [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+	err := cmd.Execute()
+	switch {
+	case err == nil:
+		// Ignore this.
+	case errResult != "":
+		assert.Equal(t, errResult, err.Error())
+		return
+	default:
+		t.Fatal(err)
+	}
```
Comment (jimeh) [inline cmd/katapult/error_utils.go:25]: I think we can simplify this switch case to a single if condition, check my comment on `Test_newClient` for reference.

File: cmd/katapult/data_centers_test.go [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+	return m.dcs, nil, nil
+}
+
+func (m mockDataCentersClient) Get(
+	_ context.Context, ref core.DataCenterRef) (*core.DataCenter, *katapult.Response, error) {
+	if m.throws != "" {
+		return nil, nil, errors.New(m.throws)
+	}
+	for _, v := range m.dcs {
+		if v.ID == ref.Permalink {
```
Comment (jimeh) [inline cmd/katapult/data_centers_test.go:54]: I think this should compare against `v.Permalink`, not `v.ID`. Additionally the IDs in the data center fixtures should be `dc_` followed by a 16 character long random alphanumeric string, for example: `dc_9UVoPiUQoI1cqtR0`

**Review by jimeh — CHANGES_REQUESTED** [review]
Apologies for the amount of comments :S... Most are nitpicks however, and the ones that aren't are things I managed to miss last week, sorry :(

File: cmd/katapult/data_centers_test.go [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+	cmd := dataCentersCmd(mockDataCentersClient{dcs: fixtureDataCenters})
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			stdout := &bytes.Buffer{}
+			stderr := &bytes.Buffer{}
+			cmd.SetOut(stdout)
+			cmd.SetErr(stderr)
+			cmd.SetArgs(tt.args)
+			err := cmd.Execute()
+			assert.Equal(t, tt.stderr, stderr.String())
```
Comment (jimeh) [inline cmd/katapult/data_centers_test.go:148]: Unless I'm blind, I think `tt.want` is never asserted against `stdout` in this test. Also I think we could use the `testAssertCommand()` helper here like in networks and orgs tests :)

File: cmd/katapult/main_test.go [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"bytes"
+	"testing"
+
+	"github.com/spf13/cobra"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func testAssertCommand(t *testing.T, cmd *cobra.Command, errResult, want, stderrResult string) {
```
Comment (jimeh) [inline cmd/katapult/main_test.go:12]: I know I suggested the name `testAssertCommand`, but the more I'm reading it now, it makes me think it's testing an "assertion command" 🤦‍♂️... So what do you think about calling it something like `assertCobraCommand` to make it more specific about it's purpose and function?

Along the lines of making things clearer, what do you think about renaming the three string arguments to `wantErr`, `wantStdout`, and `wantStderr`?

And finally, I think we should call `t.Helper()` at the beginning of the function to denote it as a helper function.

File: cmd/katapult/config_test.go [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+				expectedAPIKey = "\"\""
+			}
+			expectedAPIURL := tt.apiURL
+			if expectedAPIURL == "" {
+				expectedAPIURL = "\"\""
+			}
+			mockOutput := fmt.Sprintf(mockConfigFormat, expectedAPIKey, expectedAPIURL)
+			if out.String() != mockOutput {
+				t.Fatal("invalid result:\n\n", out.String())
+			}
```
Comment (jimeh) [inline cmd/katapult/config_test.go:69]: I'd suggest replacing this with a `assert.Equal(...)`.

File: cmd/katapult/config_test.go [hunk 32]
```diff
@@ -0,0 +1,72 @@
...
+			conf, err := config.New()
+			assert.NoError(t, err)
+			conf.SetDefault("api_key", tt.apiKey)
+			conf.SetDefault("api_url", tt.apiURL)
+			cmd := configCommand(conf)
+			out := &bytes.Buffer{}
+			cmd.SetOut(out)
+			if err := cmd.RunE(cmd, []string{}); err != nil {
+				t.Fatal(err)
+			}
```
Comment (jimeh) [inline cmd/katapult/config_test.go:57]: I'd suggest replacing this a `require.NoError(...)`.

Comment (jimeh) [reply cmd/katapult/config_test.go:57]: Actually, any reason this test can't or shouldn't use the `testAssertCommand` helper function?

**Review by jimeh — CHANGES_REQUESTED** [review]

**Review by jimeh — APPROVED** [review]
👍
